k3b: separate wrapper derivation to easily override cdrtools (e.g. with cdrkit)

Yarny0 opened the pull request:
This commit replaces the dependency on
cdrtools with a dependency on cdrkit.
The change affects the PATH seen by k3b (set by a wrapper);
k3b finds the required binaries by itself.

Advantage of cdrkit:
Based on experiment (by committer),
it permits writing optical media without superuser privileges.

Advantage of cdrtools:
It seems well supported and allows writing blu-ray media
(at least that's what the k3b feature list claims).

This commit should fix
https://github.com/NixOS/nixpkgs/issues/19154 .

###### Motivation for this change

Writing optical media should work without superuser privileges.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`): k3b fails with same error that also occurs without this commit, which is probably due to NixOS 18.03 environment: `qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""`
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @Phreedom @svanderburg

coretemp commented:
In my experience cdrtools is a superior tool to cdrkit for writing DVDs and CDs, so I would like to *not* see this merged. 

In fact, I think cdrkit wasted a couple of CDs and DVDs whereas cdrtools has never failed. 

Having both on PATH is a reasonable option (disk space is cheap, wasted time on a broken CD is not). 

If it uses cdrkit for Blueray only (if that's the only option) then that's fine, but let's not throw out a program that actually works (cdrtools).

Yarny0 committed:
k3b: wrapper as separate derivation

The k3b package uses a wrapper script to inject binaries
that are called by k3b and do the actual work into PATH.
Most of those binaries are beyond reproach.
There are, however, legitimate reasons to replace
cdrtools with cdrkit, or with nothing at all to encourage
k3b to use suid wrappers from the surrounding PATH.

To ease the use of such cdrtools replacements,
this commit moves the wrapper script into its
own derivation (created with symlinkJoin).
To use k3b with cdrkit, this line in configuration.nix
suffices and won't trigger a rebuild of k3b itself:
> environment.systemPackages = with pkgs; [ k3b.override { cdrtools = cdrkit; } ];
Using "null" for cdrtools is possible and will skip
the addition of cdrtools to PATH in the wrapper script.

This commit is motivated by
https://github.com/NixOS/nixpkgs/issues/19154 .

Yarny0 commented:
@coretemp That's interesting .. do you run cdrtools as root or at least with special capabilities (k3b as root, or cdrecord as setuid or modified with setcap)?

Given your observation, I see three possible scenarios here:
* Run k3b with cdrtools installed with special capabilities: This seems to be the official way to do it, see http://cdrtools.sourceforge.net/private/man/cdrecord/cdrecord.1.html or https://negativo17.org/cdrtools/ .
* Run k3b with cdrtools without any special privileges: this fails on my machine
* Run k3b with cdrkit: works on my machine

I hesitate to add both cdrtools and cdrkit to the PATH: k3b lets the user decide which to use, but it stores this choice as an absolute path (/nix/store/123abc..-cdrtools/bin/cdrecord) in its configuration file. Therefore I expect k3b to get confused if those paths change due to an update.

I'm wondering if it makes sense to create three k3b derivations for the scenarios above, or one without a wrapper that can easily be extended with a wrapper containing cdrtools or cdrkit or maybe nothing so that k3b finds wrappers in /run/wrappers/bin . Would that be a good idea?

Yarny0 commented:
I'm closing this one for now as the approach seems to be fruitless. As soon as I find some spare time I will try to propose a solution that encompasses all scenarios (that I know of).
